engine/runtime/src/main/scala/org/enso/interpreter/instrument/job/UpsertVisualisationJob.scala

```diff
@@ -390,6 +438,55 @@ object UpsertVisualisationJob {
     }
   }
 
+  /** Invalidate the first cached dependent node of the provided expression.
+    *
+    * @param expressionId the expression id
+    */
+  private def invalidateFirstDependent(
+    expressionId: ExpressionId
+  )(implicit ctx: RuntimeContext): Unit = {
+    val module = ctx.executionService.getContext
+      .findModuleByExpressionId(expressionId)
+      .orElseThrow(() => new ModuleNotFoundForExpressionException(expressionId))
+
+    module.getIr
+      .getMetadata(DataflowAnalysis)
+      .foreach { metadata =>
+        module.getIr.preorder
+          .find(_.getExternalId.contains(expressionId))
+          .collect {
+            case name: IR.Name =>
+              DataflowAnalysis.DependencyInfo.Type
+                .Dynamic(name.name, Some(expressionId))
+            case ir =>
+              DataflowAnalysis.DependencyInfo.Type
+                .Static(ir.getId, ir.getExternalId)
+          }
+          .flatMap { expressionKey =>
+            metadata.dependents.getExternal(expressionKey)
+          }
+          .foreach { dependents =>
+            val stacks = ctx.contextManager.getAllContexts.values
+            stacks.foreach { stack =>
+              stack.headOption.foreach { frame =>
+                dependents
```

Comment:
Invalidates caches for all the dependent values.

Are the  `dependents` transitive or direct? If they are direct, wouldn't `x.y.z.inc` cause a problem? If they are transitive will the invalidation stop on _"assignment"_ - e.g. not propagate to next statement which may also depend on the changed expression?

Reply:
`dependents` are transitive. There is another function `getExternalDirect` to get the direct dependencies